### PR TITLE
Add Scalar/Datum abstraction (#1047)

### LINKED ISSUE #1047: Add `Scalar` / `Datum` support to compute kernels
**Is your feature request related to a problem or challenge? Please describe what you are trying to do.**

When implementing analytics, users often want to do operations like `array` + `constant` or `array + array`

The Rust implementation of Arrow is strongly typed 🙌  , including the [compute kernels](https://docs.rs/arrow/6.4.0/arrow/compute/kernels/index.html), but often data is passed around in `Arc<dyn Array>` / `ArrayRef` so the types can be dynamic. This then places the burden on the user of the library to determine the exact types of the inputs in order to call the appropriate kernels

Let's take for example, trying to compare an array to find all elements that contain the number `5` and your input an array may have `Int64` or `Int32` values

In the current version of arrow, in order to do so you would need to write code like

```rust
use arrow::compute::eq_scalar;

fn find_5(array: ArrayRef) -> Result<BooleanArray>{
  match array.data_type() {
    DataType::Int64 => eq_scalar(array.as_any().downcast_ref::<Int64Array>().unwrap(), 5),
    DataType::UInt64 => eq_scalar(array.as_any().downcast_ref::<UInt64Array>().unwrap(), 5),
    ...
  }
}
```
This ends up being macroized and is non ideal because the user had to do dynamic type dispatch anyways, so there is no runtime performance benefit to strongly typed kernels

**Describe the solution you'd like**

It would be nice to be able to call a single function and let the rust library dynamically (at runtime) pick the correct kernel to call.

As described and suggested by @jorgecarleitao  and @nevi-me  in https://github.com/apache/arrow-rs/pull/984#issuecomment-994093090 this could go all the way and take the form of following the C++ `Datum` model

```rust
fn eq(left: Datum, right: Datum) -> BooleanArray {
...
}
```

Where `Datum` looks something like

```rust
enum Datum {
  Scalar(ScalarValue),
  Array(ArrayRef)
}
```

And `ScalarValue` which would look something like `ScalarValue` from DataFusion https://github.com/apache/arrow-datafusion/blob/master/datafusion/src/scalar.rs or `Scalar` from arrow2 https://github.com/jorgecarleitao/arrow2/tree/main/src/scalar

**Describe alternatives you've considered**
The alternative is a continued proliferation of individual kernels such as `eq`, `eq_utf8`, etc

**Additional context**
There is a lot of additional discussion on https://github.com/apache/arrow-rs/pull/984

cc @matthewmturner @Dandandan @Jimexist @houqp

### PR DESCRIPTION
**Pull request by tustvold**
# Which issue does this PR close?

Closes #1047
Relates to #3999
Relates to #2837 
Relates to #2766

# Rationale for this change
 

This is a proposal on how to represent scalars within arrow-rs kernels, with a view to providing a more consistent experience for users, and allowing us to consolidate a non-trivial amount of dispatch logic currently found in [downstreams](https://github.com/apache/arrow-datafusion/blob/d9e91d187c8af7f39999f8b1a11d53383826a471/datafusion/physical-expr/src/expressions/binary.rs).

There are a couple of aspects worth highlighting

## Arrays are Preeminent

Arrow is a specification for arrays and not scalars. As such this approach makes a conscious decision to not define a parallel specification for representing scalars. This has a number of advantages:

* Can use the existing stories for serialization, FFI, display, casting, etc...
* Avoids ambiguity over how to represent scalars with non-native encodings, e.g. dictionaries, run-end encoded, etc...
* Allows sharing more logic, by not bifurcating the types
* Supports all array types with no additional effort

It does come with some obvious downsides compared to a first-class scalar approach

* Less efficient representation
* Potentially more confusing / less ergonomic to use

However, it is my assertion that preserving the arrow-ness of the representation is more important than either of these

## Non-Owning

Related to the above, data is **only** stored in arrays. `Datum` and `Scalar` solely act as wrappers to influence the dispatch within kernels, **you cannot store a scalar value**. Similarly the return type of a kernel will always be an array.

This reflects both the desire to keep arrays as the canonical representation, and also to discourage use of the scalar abstraction within data structures. We want to encourage the use of arrays, not constructions like `Vec<Scalar>` or `HashMap<Scalar, _>`. Permitting such usage not only creates confusion, but performing type-erasure per field in this way is grossly inefficient both from a memory and performance standpoint.

## Type-Erasure / Dynamic Dispatch

Both `Scalar` and `Datum` are abstractions relying on type-erasure. This reflects a couple of goals:

* Designing generics for many operations, like arithmetic, is extremely hard and represents a significant API commitment, type-erasure presents a much smaller API surface that is therefore easier to evolve without breaking changes
* Avoiding generics helps to reduce code bloat from monomorphisation and hopefully improve compile times
* Is consistent with the vast majority of kernels which use `&dyn Array`

The implied assumption is that the overheads of this dynamic dispatch will be irrelevant when amortised over the number of values in an array. Or to phrase it differently, **we are explicitly not optimising for performance of purely scalar operations**. IMO such operations are outside the remit of a vectorised execution engine.

# What changes are included in this PR?

Adds `Scalar` and `Datum` along with examples of their usage

# Are there any user-facing changes?

No, this just adds the new types

### CODE REVIEW

File: arrow-array/src/array/boolean_array.rs [hunk 1]
```diff
@@ -93,6 +93,17 @@ impl BooleanArray {
         Self { values, nulls }
     }
 
+    /// Create a new [`BooleanArray`] with length `len` consisting only of nulls
+    pub fn new_null(len: usize) -> Self {
```
Comment (tustvold) [inline arrow-array/src/array/boolean_array.rs:97]: This was a change to make the example work

File: arrow-select/src/filter.rs [hunk 2]
```diff
@@ -321,16 +321,6 @@ fn filter_array(
         // actually filter
         _ => downcast_primitive_array! {
             values => Ok(Arc::new(filter_primitive(values, predicate))),
-            DataType::Decimal128(p, s) => {
```
Comment (tustvold) [inline arrow-select/src/filter.rs:324]: Drive-by fix I noticed whilst implementing this, downcast_primitive_array already handles the decimal case.

File: arrow-array/src/scalar.rs [hunk 3]
```diff
@@ -0,0 +1,116 @@
...
+/// # use arrow_schema::ArrowError;
+/// #
+/// fn eq_impl<T: ArrowPrimitiveType>(
+///     a: &PrimitiveArray<T>,
+///     a_scalar: bool,
+///     b: &PrimitiveArray<T>,
+///     b_scalar: bool,
+/// ) -> BooleanArray {
+///     let (array, scalar) = match (a_scalar, b_scalar) {
+///         (true, true) | (false, false) => {
```
Comment (tustvold) [inline arrow-array/src/scalar.rs:41]: Here we can see a nice property of the encoding as scalars, the case where they are both scalars is identical to the case where they are both arrays

**Comment by tustvold** [comment]
Mailing list thread - https://lists.apache.org/thread/f5q4q1g6hgqgrvbv67llv87hpfxg34vc

File: arrow-select/src/filter.rs (hunk 2 above)
Comment (alamb) [reply arrow-select/src/filter.rs:324]: might make sense to put up as a new PR

**Review by alamb — COMMENTED** [review]
TLDR is I think this is really nice -- thank you @tustvold 

BTW for anyone else reading this, I think one of the major reasons @tustvold  is tackling this now is so that as we add scalar kernels for the various timestamp / duration / interval types we don't make the existing problem worse. 

I focused on this example, which shows how someone would use the kernels with scalar values:

```rust
/// // Comparison of an array and a scalar
/// let a = Int32Array::from(vec![1, 2, 3, 4, 5]);
/// let b = Int32Array::from(vec![1]);
/// let r = eq(&a, &Scalar::new(&b)).unwrap();
/// let values: Vec<_> = r.values().iter().collect();
/// assert_eq!(values, &[true, false, false, false, false]);
```

I really like the construction of a `Scalar` that wraps `b` and signals to the kernels that the 1 element array should be treated differently, rather than just implicitly treating 1 element arrays differently

I also think the use of a trait for `Datum` is important for uses like DataFusion, which will likely need some sort of owned variant of Scalar (e.g. to represent the `1` in an expression such as `column + 1`). Given that `Datum` is a trait means we can create such an abstraction. 

It might make sense to add an `OwnedScalar` in arrow-rs for convenience (`OwnedScalar(ArrayRef)`) but we can sort that out later 

Questions:
Have we thought about how we will migrate the existing kernels? Given the construction above, we could perhaps leave the old signatures around for a while and deprecate them and call through to the new variants

File: arrow-array/src/scalar.rs [hunk 4]
```diff
@@ -0,0 +1,110 @@
...
+///
+/// // Comparison of an array and a scalar
+/// let a = Int32Array::from(vec![1, 2, 3, 4, 5]);
+/// let b = Int32Array::from(vec![1]);
+/// let r = eq(&a, &Scalar::new(&b)).unwrap();
+/// let values: Vec<_> = r.values().iter().collect();
+/// assert_eq!(values, &[true, false, false, false, false]);
+pub trait Datum {
+    /// Returns the value for this [`Datum`] and a boolean indicating if the value is scalar
+    fn get(&self) -> (&dyn Array, bool);
```
Comment (alamb) [inline arrow-array/src/scalar.rs:80]: It might make sense to have two separate methods so that if we wanted to add more information to the Datum we could do it in a backwards compatible way by adding a new method to the trait

```rust
fn get(&self) -> &dyn Array;
fn is_scalar(&self) -> bool {
 false 
}
```

**Comment by tustvold** [comment]
> Have we thought about how we will migrate the existing kernels

My hope is that we can simply replace `&dyn Array` with `&dyn Datum` and have everything work without it requiring downstreams to modify their code, but I haven't tested this extensively yet

File: arrow-array/src/scalar.rs (hunk 4 above)
Comment (tustvold) [reply arrow-array/src/scalar.rs:80]: I think given that neither piece of information can be used in isolation, it makes sense to return them both together. We can still always add further trait methods to expose additional information in future

File: arrow-array/src/array/boolean_array.rs (hunk 1 above)
Comment (tustvold) [reply arrow-array/src/array/boolean_array.rs:97]: Split out into - https://github.com/apache/arrow-rs/pull/4402

**Comment by tustvold** [comment]
Another thing to potentially consider at the same time is whether we want to roll in any notion of selection vector support (#4095). I will think on this more

**Comment by tustvold** [comment]
https://github.com/apache/arrow-rs/pull/4465 contains a POC of using this abstraction to implement scalar kernels, I'm therefore confident that this is a sensible abstraction

**Review by alamb — APPROVED** [review]

File: arrow-array/src/scalar.rs [hunk 5]
```diff
@@ -0,0 +1,116 @@
...
+}
+
+impl<T: Array> Datum for T {
+    fn get(&self) -> (&dyn Array, bool) {
+        (self, false)
+    }
+}
```
Comment (viirya) [inline arrow-array/src/scalar.rs:87]: Hmm, so an Array with length 1 will also return false indicating it is not a scalar? Only if we explicitly wrap it with a `Scalar` so we can get correct output?

Comment (tustvold) [reply arrow-array/src/scalar.rs:87]: Yes, it was suggested this might be less confusing for users

**Comment by tustvold** [comment]
I intend to merge this after I cut the arrow 43 release

File: arrow-array/src/scalar.rs (hunk 5 above)
Comment (tustvold) [reply arrow-array/src/scalar.rs:87]: https://github.com/apache/arrow-rs/pull/4393/files#diff-7174f433d1c74c0e47944054399558364f11d0a813e787b854196467fc7d7f8bR114

?

Comment (viirya) [reply arrow-array/src/scalar.rs:87]: That is Datum for Scalar, I mean Datum for dyn Array, then so an Array with length 1 will be treated as scalar without Scalar?

File: arrow-array/src/scalar.rs [hunk 6]
```diff
@@ -0,0 +1,116 @@
...
+    /// Panics if `array.len() != 1`
+    pub fn new(array: &'a dyn Array) -> Self {
+        assert_eq!(array.len(), 1);
+        Self(array)
+    }
+}
+
+impl<'a> Datum for Scalar<'a> {
+    fn get(&self) -> (&dyn Array, bool) {
+        (self.0, true)
```
Comment (viirya) [inline arrow-array/src/scalar.rs:114]: Do we need to consider datatype of the wrapped Array? A complex type array with length 1 is also a scalar?

File: arrow-array/src/scalar.rs (hunk 5 above)
Comment (tustvold) [reply arrow-array/src/scalar.rs:87]: An array with length 1 won't be treated as scalar, it will only be treated as scalar if wrapped in `Scalar`?

File: arrow-array/src/scalar.rs (hunk 6 above)
Comment (tustvold) [reply arrow-array/src/scalar.rs:114]: If wrapped in `Scalar` then yes, this is consistent with DF's ScalarValue which can contain struct or list elements

File: arrow-array/src/scalar.rs (hunk 5 above)
Comment (viirya) [reply arrow-array/src/scalar.rs:87]: Yea, that is what I wanted to ask, I think. Not all arrays with length 1 are all scalars, but only wrapped with `Scalar` they are scalars.

The view maybe be reversed. Not array is treated as scalar but scalar is treated as array. When any one wants to talk with this crate, this crate only understands language in array. So if you want to mention a scalar, you need to fit it into an array and let this crate know it behaves like a scalar.

Comment (viirya) [reply arrow-array/src/scalar.rs:87]: There are some downsides, but I think you already mentioned in the description.

Comment (alamb) [reply arrow-array/src/scalar.rs:87]: > An array with length 1 won't be treated as scalar, it will only be treated as scalar if wrapped in Scalar

FWIW I think the only practical difference would be that `add(arr1, arr2)` will fail if `arr1` has one row (but is not marked as a scalar) and `arr2` had some other number of rows (like 100). 

I think @tustvold  also considered simply treating any arrays that had `1` row as a scalar but felt (as do I) that making it explicit would make for a less confusing experience . Or maybe that was only my opinion 😆 

> When any one wants to talk with this crate, this crate only understands language in array. So if you want to mention a scalar, you need to fit it into an array and let this crate know it behaves like a scalar.

I think this is an excellent description 👍

**Review by viirya — APPROVED** [review]
